run_tests.py

```diff
@@ -80,24 +83,47 @@ class GRREverythingTestLoader(test_lib.GRRTestLoader):
   base_class = test_lib.GRRBaseTest
 
 
+class GRROneTestLoader(test_lib.GRRTestLoader):
+  """Load one specific GRR test case."""
+  base_class = test_lib.GRRBaseTest
+
+  def loadTestsFromModule(self, _):
+    """Just return all the tests as if they were in the same module."""
+    test_cases = [
+        self.loadTestsFromTestCase(x) for x in self.base_class.classes.values()
+        if x.__name__ == self.test_class_name]
+
+    return self.suiteClass(test_cases)
+
+  def __init__(self, test_class_name):
+    self.test_class_name = test_class_name
+
+    super(GRROneTestLoader, self).__init__()
+
+
 def RunTest(test_suite, stream=None):
   out_fd = stream
   if stream:
     out_fd = StringIO.StringIO()
 
   try:
-    test_lib.GrrTestProgram(argv=[sys.argv[0], test_suite],
-                            testLoader=GRREverythingTestLoader(
-                                labels=flags.FLAGS.labels),
-                            testRunner=unittest.TextTestRunner(
-                                stream=out_fd))
+    program = test_lib.GrrTestProgram(
+                argv=[sys.argv[0], test_suite],
+                testLoader=GRROneTestLoader(
+                    test_class_name=test_suite),
+                testRunner=unittest.TextTestRunner(
+                    stream=out_fd))
+  except SystemExit as e:
+    # must include this section, as GrrTestProgram issues
```

Comment:
> The existing test script can run in two modes - inline: That is really only designed to run one test at the time and its used for debugging mostly (i.e. so you can put break points etc). The single process mode is really designed to run interactively only.

It is very unclear to me why the single-process mode is "designed to run one test at the time". There are no comments indicating this. There is no documentation indicating this. The code is written in a way that strongly indicates otherwise. As I said in my previous comment, and as you can readily see in the code, there is a **for loop** looping through multiple tests in the single-process mode path (also quoted below). Why is there a for loop looping through an array which is only supposed to have one element by design according to what you said? Why is it even an array then and not just a single value? This doesn't make any sense. This code is just badly designed and buggy and it's not aware of the fact that there is an exit condition within the GrrProgram class. It seems that simply single-process mode with multiple tests was never used, even though the program seems designed to handle this situation, but the implementation is failing. Please [take a look at the code](https://github.com/google/grr/blob/master/run_tests.py#L177) and inspect for yourself. This for loop goes through all the suites and it's in the single process path.

This PR works around this bug for now so that we can use single-process mode to run multiple tests, which is the intention indicated by the semantics of the previous code.

> It appears Travis cant run more than one test at the time due to resource issues, so we need to reduce the number of concurrent tests to one - currently running the tests inline is not really a supported mode of operation and the code specifically checks for processes == 1 and goes into that code path.

This PR works around the bugs of single-process mode and makes it possible to run multiple tests in single-process mode.

>  You claim that this is not reliable since some tests are exiting

I'm not claiming "some tests are exiting". I'm claiming *all* tests are exiting. This is the expected behavior for Python unittest TestProgram. [Here is the relevant python standard library code patch that defines the current intentional exit behavior for Python unittest TestProgram](https://bugs.python.org/file13790/unittest-no-exit.patch) which `GrrTestProgram` extends. The previous code in multiprocessing mode only ever worked because there was a peculiar set of bugs that was making the program behave correctly. Here they are:

* **The first bug**: In the previous code, the `GrrTestProgram` is expected to **not** exit upon completion. This is why it is called within a function which is called within a `for` loop which invokes it several times. Here is the relevant code snippet that illustrates we don't expect this to exit:

```
for test_suite in suites:
  RunTest(test_suite, stream=stream)                                                                                                                
```

This code snippet clearly doesn't assume `RunTest` to exit the python process. Otherwise it wouldn't have been a `for` loop.

* **The second bug**: In the previous code, `GrrTestProgram` ignores its test_suite parameter. However, the code is not aware of this. Here is the relevant code snippet that illustrates this:

```
def RunTest(test_suite, stream=None):                                                                                                                   
  out_fd = stream

  test_lib.GrrTestProgram(argv=[sys.argv[0], test_suite],
                          testLoader=GRREverythingTestLoader(
                              labels=flags.FLAGS.labels),
                          testRunner=unittest.TextTestRunner(
                              stream=out_fd))
```

By reading this function, you can see that `test_suite` is expected to be the name of the test that we are about to run. This expectation is also clear from the signature of the function `RunTest`:

```
RunTest(test_suite, stream=None)
```

By reading this function signature and with no docstrings provided, it is reasonable to assume that this function runs the test specified in the `test_suite` parameter. Instead, the behavior is to ignore the `test_suite` parameter.

* **The third bug**: In the previous code, `GrrTestProgram` reads directly from `flags`. This behavior is unexpected and magical. Flags are incorrectly used as global variables. For the same reason [global variables are bad](http://c2.com/cgi/wiki?GlobalVariablesAreBad), this use of flags is bad and makes things hard to debug. This should be treated like a bug and fixed. Either `GrrTestProgram` must be invoked at the top-level and left to handle `flags` itself, or it should be invoked as a runner within the `RunTest` function and pay attention to its parameters without magically reading from `flags`.

In the end, the use of a `TestProgram` is not appropriate at this place at all. I will investigate more into alternatives of refactoring this, either using `nose` or by refactoring the existing use of `GrrTestProgram`.

>  so I am suggesting we stick to the multiprocess mode since it is the current way things work and will require minimal new code.

No. Multiprocess mode doesn't work on travis. I once again point you to the [relevant tests I ran](https://s3.amazonaws.com/archive.travis-ci.org/jobs/45939849/log.txt) which clearly illustrate that multiprocess mode will not work in this case. It may work with some effort, but this effort may require refactoring and is certainly not "minimal" in code changes. Single process mode, the way I worked-around it in this PR, works. I don't think the code changes made in this PR are that humongous. We can refactor later to use multiprocessing mode if you insist.

> All the grr test programs are stand alone (i.e. they have their own main()) and are actually designed to run stand alone. This is the best practice for running tests because you ensure that you have test isoltion when you start a new process each time (you can not guarantee this if you dont because we have a lot of shared globals like config library etc). In Rekall we actually spawn the test binary itself but I think in GRR we fork (maybe we should spawn).

> Not all tests should run as a separate program but test suites (which are collections of tests) should.

OK, sounds good, but this is not the intention of this PR.

> I think changing the test framework is probably outside the scope of this cl.

Agreed.

> I just suggested to remove that processes == 1 condition from the top and just see if that improves things on travis by running the test program in a child process.

This sounds like an experiment worthy exploring. Let's do that in a different PR. Could you open an issue?

> Note that tests can fail also due to a seg fault and catching the exception wont help here.

I agree in principle, but this is out of the scope of this PR. This PR's purpose is to get our tests running on travis, one way or another. Optimizing for performance or extreme cases such as segfaults can be done later. In fact, [I have verified that running all the tests without exclusions does not cause any segfaults currently](https://s3.amazonaws.com/archive.travis-ci.org/jobs/46216964/log.txt), even though it causes expected and unexpected failures which are handleable, so, while this may not be the best course of action and I see your point in desiring isolation, it's something we can live with for a while and segfaults can be regarded as non-existing extremities at this stage.